website/pages/docs/concepts/seal.mdx

```diff
@@ -84,80 +84,167 @@ to decrypt the master key at startup and unseal automatically.
 When using Auto Unseal there are certain operations in Vault that still
 require a quorum of users to perform an operation such as generating a root token.
 During the initialization process, a set of Shamir keys are generated that are called
-Recovery Keys and are used for these operations.
+recovery keys and are used for these operations.
 
 For a list of examples and supported providers, please see the
 [seal documentation](/docs/configuration/seal).
 
+## Recovery Key Rekeying
+
+During the KMS Seal initialization process, a set of Shamir keys called recovery keys are
+generated which are used for operations that still require a quorum of users.
+
+Recovery keys can be rekeyed to change the number of shares or thresholds. When using the
+Vault CLI, this is performed by using the `-target=recovery` flag to `vault operator rekey`.
+
 ## Seal Migration
 
-The seal can be migrated from Shamir keys to Auto Unseal, Auto Unseal to Shamir
-keys and Auto Unseal to another Auto Unseal.
+The seal can be migrated from Shamir Seal to KMS Seal, KMS Seal to Shamir Seal,
+and KMS Seal to another KMS Seal.
+
+~> **NOTE**: Seal migration process cannot be performed without downtime. Due to
+the intricate technical underpinnings of the seal feature, it is at this point
+not possible to perform seal migration without bringing the whole cluster down.
+We understand that it can be painful for many users to bring the full cluster
+down. But, we believe that switching seals is a very rare event and hence hope
+for the downtime to be considered as an acceptable trade-off.
+
+~> **NOTE**: Seal migration operation will require both seals to be available
+during the migration. For example, a migration from a KMS seal to Shamir will
+require that the KMS key be accessible during the migration.
+
+~> **NOTE**: Seal migration from KMS seal to Shamir seal is not currently
+supported when using Vault Enterprise. We plan to support this officially in a
+future release.
+
+~> **NOTE**: Seal migration from KMS seal to KMS seal of same kind is not
+currently supported when using Vault Enterprise. We plan to support this
+officially in a future release.
+
+### Migration From Shamir Seal to KMS Seal
+
+#### When **not** using Integrated Storage
```

Comment:
I don't understand why we cannot use one set of instructions for any storage type. Since only the active node does the migration can't we follow the same steps as integrated storage for any storage backend. I think the following steps will work for any storage:

1. Take Standby nodes down and update their configs.
2. Start the Standby nodes and unseal them with -migrate
3. Stop the active node.
4. One of the standby nodes will become active and preform the migration.
5. Depending on the migration type update the config on the old active node and start it.

This flow also minimizes downtime to the failover time + the time it takes to migrate. Which is better than having to stop all nodes at once.

Reply:
The primary reason for this approach is cleanup.

For storage backends that are not using raft, there is really no need to setup all the nodes in the cluster for migration. The first/active node fully completes the operation and standby nodes can join the cluster with no knowledge of migration (no -migrate flag or no old seal blocks required). I thought this could be easier to reason and work with. Seal migration requires the config file to be cleaned up after the migration is complete (removal of old seal blocks). Having one node perform the migration means that there is one node's config that needs cleanup. Downtime is anyways inevitable. I thought that it might be better to work with less moving parts than probably saving a few seconds of downtime.

For raft storage, migration can be performed by quorum number of nodes. Then cleanup of all the quorum number of nodes' config file through failover/update approach. After this, addition of the rest of the peer nodes will be with no knowledge of migration.

Reply:
I am okay to go either ways. Just wanted to let you know the thought process.

Reply:
> Downtime is anyways inevitable. I thought that it might be better to work with less moving parts than probably saving a few seconds of downtime.

I'm not sure this is true, with the steps I outlined above the only disruption you'd have is during the failover from active to standby. This is much less than shutting all nodes down, updating a config file, and starting them up. 

> For storage backends that are not using raft, there is really no need to setup all the nodes in the cluster for migration.

It depends on what you're migrating to/from. You may want to include the migration seal on all nodes incase you need to unseal-wrap a lot of data. 

Plus I think it's simpler to have one set of instructions than to recommend different practices based on your storage backend. 

I also would like to see a pre-1.4 set of steps and a post-1.4 set of steps since the approach changed quite a bit in 1.4.

Reply:
@briankassouf Let me know what you think of the updates.